sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -376,35 +385,76 @@ def get(self, instruction_id, bundle_descriptor_id):
 
     Moves the ``BundleProcessor`` from the inactive to the active cache.
     """
-    try:
-      # pop() is threadsafe
-      processor = self.cached_bundle_processors[bundle_descriptor_id].pop()
-    except IndexError:
-      processor = bundle_processor.BundleProcessor(
-          self.fns[bundle_descriptor_id],
-          self.state_handler_factory.create_state_handler(
-              self.fns[bundle_descriptor_id].state_api_service_descriptor),
-          self.data_channel_factory)
-    self.active_bundle_processors[
```

Comment:
I'm not sure what change your suggesting.

Note, it is important to not hold the lock while we instantiate the BundleProcessor or access `self.fns`

Reply:
isn't it the same if we do
```
    try:
      # pop() is threadsafe
      processor = self.cached_bundle_processors[bundle_descriptor_id].pop()
    except IndexError:
      processor = bundle_processor.BundleProcessor(
          self.fns[bundle_descriptor_id],
          self.state_handler_factory.create_state_handler(
              self.fns[bundle_descriptor_id].state_api_service_descriptor),
          self.data_channel_factory)

    with self._lock:
      self.active_bundle_processors[
        instruction_id] = bundle_descriptor_id, processor
      try:
        del self.known_not_running_instruction_ids[instruction_id]
      except KeyError:
        # The instruction may have not been pre-registered before execution
        # since activate() may have never been invoked
        pass
    return processor
```

Reply:
The `pop` is thread safe but accessing `self.cached_bundle_processors[bundle_descriptor_id]` is not guaranteed to be so your right that some could be saved based upon your suggestion but it still requires a self.lock surrounding
```
processor = self.cached_bundle_processors[bundle_descriptor_id].pop()
```

which will lead to us acquiring the mutex twice for the case when there is a cached instance while the current imlementation only needs the mutex twice if we need to create the BundleProcessor.